src/datasets/arrow_dataset.py

```diff
@@ -1707,46 +1710,60 @@ def format_cache_file_name(cache_file_name, rank):
+                    offset=sum(len(s) for s in shards[:rank]),
+                    disable_tqdm=disable_tqdm,
+                    desc=desc,
+                )
+                for rank in range(num_proc)
+            ]
+
+            # Upon mapping using multiple processes, we can choose to load them on a single process instead by doing it
+            # sequentially. This is useful when preprocessing can be done on multiple processes and loading has to be
+            # done on a single process.
+            if sequential:
+                logger.warning(
+                    "Running map on shards sequentially, despite having `num_proc={}`. This is generally used to load a"
+                    " multiprocessed cached mapping.".format(num_proc)
+                )
                 os.environ = prev_env
-                shards = [
-                    self.shard(num_shards=num_proc, index=rank, contiguous=True, keep_in_memory=keep_in_memory)
-                    for rank in range(num_proc)
-                ]
-                kwds_per_shard = [
-                    dict(
-                        self=shards[rank],
-                        function=function,
-                        with_indices=with_indices,
-                        input_columns=input_columns,
-                        batched=batched,
-                        batch_size=batch_size,
-                        drop_last_batch=drop_last_batch,
-                        remove_columns=remove_columns,
-                        keep_in_memory=keep_in_memory,
-                        load_from_cache_file=load_from_cache_file,
-                        cache_file_name=format_cache_file_name(cache_file_name, rank)
-                        if cache_file_name is not None
-                        else None,
-                        writer_batch_size=writer_batch_size,
-                        features=features.copy() if features is not None else None,
-                        disable_nullable=disable_nullable,
-                        fn_kwargs=fn_kwargs,
-                        rank=rank,
-                        offset=sum(len(s) for s in shards[:rank]),
-                        disable_tqdm=disable_tqdm,
-                        desc=desc,
-                    )
-                    for rank in range(num_proc)
-                ]
-                logger.info("Spawning {} processes".format(num_proc))
-                results = [pool.apply_async(self.__class__._map_single, kwds=kwds) for kwds in kwds_per_shard]
-                transformed_shards = [r.get() for r in results]
-                logger.info("Concatenating {} shards from multiprocessing".format(num_proc))
-                result = concatenate_datasets(transformed_shards)
-                if new_fingerprint is not None:
-                    result._fingerprint = new_fingerprint
-                return result
+                transformed_shards = [self.__class__._map_single(**dill.copy(kwds)) for kwds in kwds_per_shard]
```

Comment:
`diil.copy`: This is needed in order to mimick multiprocessing pickling. The issue is if we pass arguments using a reference, then modification get propagated through the sequence. Typically, using a random generator.